echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/eventhandlers/ManualEventHandler.java

```diff
@@ -74,25 +92,58 @@ private boolean pipelineMatches(String application, String nameOrId, Pipeline pi
       && (pipeline.getName().equals(nameOrId) || pipeline.getId().equals(nameOrId));
   }
 
-  private Pipeline buildTrigger(Pipeline pipeline, Trigger manualTrigger) {
+  protected Pipeline buildTrigger(Pipeline pipeline, Trigger manualTrigger) {
     List<Map<String, Object>> notifications = buildNotifications(pipeline.getNotifications(), manualTrigger.getNotifications());
     Trigger trigger = manualTrigger.atPropagateAuth(true);
-    List<Artifact> artifacts = Collections.emptyList();
+    List<Artifact> artifacts = new ArrayList<>();
     String master = manualTrigger.getMaster();
     String job = manualTrigger.getJob();
     if (buildInfoService.isPresent() && StringUtils.isNoneEmpty(master, job)) {
       BuildEvent buildEvent = buildInfoService.get().getBuildEvent(master, job, manualTrigger.getBuildNumber());
       trigger = trigger
         .withBuildInfo(buildInfoService.get().getBuildInfo(buildEvent))
         .withProperties(buildInfoService.get().getProperties(buildEvent, manualTrigger.getPropertyFile()));
-      artifacts = buildInfoService.get().getArtifactsFromBuildEvent(buildEvent, manualTrigger);
+      artifacts.addAll(buildInfoService.get().getArtifactsFromBuildEvent(buildEvent, manualTrigger));
+    }
+
+    if (artifactInfoService.isPresent() && manualTrigger.getArtifacts().size() != 0) {
+      List<Artifact> resolvedArtifacts = resolveArtifacts(manualTrigger.getArtifacts());
+      artifacts.addAll(resolvedArtifacts);
+      // Remove the artifact lookup information. It will be replaced with the full artifacts if they can be resolved.
+      trigger = trigger.withArtifacts(Collections.emptyList());
     }
     return pipeline
       .withTrigger(trigger)
       .withNotifications(notifications)
       .withReceivedArtifacts(artifacts);
   }
 
+  protected List<Artifact> resolveArtifacts(List<Map<String, Object>> manualTriggerArtifacts) {
+    List<Artifact> resolvedArtifacts = new ArrayList<>();
+    for (Map a : manualTriggerArtifacts) {
+      Artifact artifact =  objectMapper.convertValue(a, Artifact.class);
+
+      if (Strings.isNullOrEmpty(artifact.getName()) ||
+        Strings.isNullOrEmpty(artifact.getVersion()) ||
+        Strings.isNullOrEmpty(artifact.getLocation())) {
+        log.error("Artifact does not have enough information to fetch. " +
+          "Artifact must contain name, version, and location.");
+      } else {
+        try {
+          Artifact resolvedArtifact = artifactInfoService.get()
```

Comment:
What happens here if I send along a fully-qualified artifact already?  It looks like we won't try to resolve it (as it won't have `location` set) but that we will remove it from the received artifacts.  Should the behavior in that case be to send it along unmodified?

Reply:
Ya, that probably isn't good. I'll change it so that if we can resolve the artifact we replace it, otherwise we leave it alone.

Reply:
Fixed to send along unchanged.